### PR TITLE
Exposes Identity Center accounts as Apps in Unified Resource Cache

### PR DESCRIPTION
**Pull request by tcsc**
For the purposes of the UI, Identity Center accounts and account
assignments are treated like special Apps. This patch exposes
Account Assignments to the UI via the Unified Resource Cache.

Includes:
- Generating an App resource from an Identity Center Account resource
- General plumbing from backend through to cache and UI

### CODE REVIEW

File: api/types/role.go [hunk 1]
```diff
@@ -1913,6 +1913,13 @@ func (r *RoleV6) GetLabelMatchers(rct RoleConditionType, kind string) (LabelMatc
 		return LabelMatchers{cond.WindowsDesktopLabels, cond.WindowsDesktopLabelsExpression}, nil
 	case KindUserGroup:
 		return LabelMatchers{cond.GroupLabels, cond.GroupLabelsExpression}, nil
+	case KindIdentityCenterAccount:
+		var matchers LabelMatchers
+		accounts := utils.Transform(cond.AccountAssignments, IdentityCenterAccountAssignment.GetAccount)
+		if len(accounts) > 0 {
+			matchers.Labels = Labels{IdentityCenterAccountLabel: utils.Deduplicate(accounts)}
+		}
+		return matchers, nil
```
Comment (r0mant) [inline api/types/role.go:1922]: Do we really need a new Transform function and templating here?

TBH I've been staring at this for 10 minutes trying to understand what this is doing - wouldn't just having a simple `GetICAccounts()` function on the role conditions object that returns a list of accounts names from AccountAssignments be the equivalent?

```go
case KindIdentityCenterAccount:
    return LabelMatchers{
        Labels: Labels{
            IdentityCenterAccountLabel: utils.Deduplicate(cond.GetICAccounts()),
        }
    }
```

File: api/utils/slices.go [hunk 2]
```diff
@@ -129,3 +129,17 @@ func CountBy[S ~[]E, E any](elements S, mapper func(E) string) map[string]int {
 	}
 	return out
 }
+
+// Transform transforms a slice of values into a new slice, with values being converted by the
+// supplied mapping function
+func Transform[S ~[]T, T any, U any](src S, mapper func(T) U) []U {
```
Comment (r0mant) [inline api/utils/slices.go:135]: As mentioned above, I think we can simplify and do away with this function.

File: lib/auth/auth_with_roles.go [hunk 3]
```diff
@@ -1352,6 +1352,16 @@ func (a *ServerWithRoles) ListUnifiedResources(ctx context.Context, req *proto.L
...
 		}
 
+		// IdentityCenter resources are bundled under the umbrella KindIdentityCenter
+		// so we need to check that as well as the actual concrete kind
+		if kind == types.KindIdentityCenterAccount {
+			err := a.action(apidefaults.Namespace, types.KindIdentityCenter, actionVerbs...)
+			if err != nil {
+				resourceAccess.kindAccessMap[kind] = err
+				continue
+			}
```
Comment (r0mant) [inline lib/auth/auth_with_roles.go:1362]: Nit: Isn't this the equivalent?
```suggestion
			resourceAccess.kindAccessMap[kind] = a.action(apidefaults.Namespace, types.KindIdentityCenter, actionVerbs...)
```

File: lib/auth/auth_with_roles.go [hunk 4]
```diff
@@ -1822,9 +1832,12 @@ func (r resourceChecker) CanAccess(resource types.Resource) error {
 		}
 	case types.SAMLIdPServiceProvider:
 		return r.CheckAccess(rr, state)
+
+	case services.UnifiedResource153Adapter[services.IdentityCenterAccount]:
```
Comment (r0mant) [inline lib/auth/auth_with_roles.go:1836]: Why is this different from other cases above i.e. why aren't we matching on `types.KindIdentityCenterAccount` here?

File: lib/services/unified_resource.go [hunk 5]
```diff
@@ -670,7 +701,20 @@ func (c *UnifiedResourceCache) processEventsAndUpdateCurrent(ctx context.Context
 		case types.OpDelete:
 			c.deleteLocked(event.Resource)
 		case types.OpPut:
-			c.putLocked(event.Resource.(resource))
+			r := event.Resource
+			if u, ok := r.(types.Resource153Unwrapper); ok {
+				switch unwrapped := u.Unwrap().(type) {
+				case IdentityCenterAccount:
+					r = WrapUnifiedResource153(unwrapped)
```
Comment (r0mant) [inline lib/services/unified_resource.go:708]: This is confusing. Why do we need to unwrap and then wrap it back? Is there a cleaner way to do this?

File: lib/services/unified_resource_adapter.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils"
+)
+
+// UnifiedResource153 is a type constraint that requires a type to implement
+// the Resource153 interface AND provides a Clone method.
+type UnifiedResource153[T interface{ CloneResource() T }] interface {
+	types.Resource153
+	CloneResource() T
+}
```
Comment (r0mant) [inline lib/services/unified_resource_adapter.go:18]: Why is this needed? Is there a way to avoid another templated interface type? Can't types.Resource153 just include Clone method?

File: lib/web/ui/app.go [hunk 7]
```diff
@@ -34,6 +35,9 @@ import (
 type App struct {
 	// Kind is the kind of resource. Used to parse which kind in a list of unified resources in the UI
 	Kind string `json:"kind"`
+	// Subkind is the subkind of the app resource. Used to differentiate different
+	// flavors of app.
+	Subkind string `json:"subkind,omitempty"`
```
Comment (r0mant) [inline lib/web/ui/app.go:40]:
```suggestion
	SubKind string `json:"subkind,omitempty"`
```

**Review by r0mant — CHANGES_REQUESTED** [review]
I think this would benefit from a review from someone on the scale team as well who is familiar with unified resources cache. @rosstimothy Who would be best to take a look?

@tcsc Also, I think we're getting overly excited about generics and iterators and should dial back. It is my strong opinion that they hurt readability a lot and we should be using them _very_ sparingly and only if there's a really good reason. I think it took me 3x more time to review this than it should have because of trying to comprehend what all these templated types do and why are they needed and I strongly suspect we do not need any of them.

File: lib/services/unified_resource_adapter.go (hunk 6 above)
Comment (tcsc) [reply lib/services/unified_resource_adapter.go:18]: @r0mant, @rosstimothy: It will also need `ResourceWithLabels`, but sure. 

The whole `UnifiedResource153Adapter` exists because I was trying to get a RFD153-style Resource into the Unified Resource Cache while still following follow the advice in  the [types.Resource153 godoc](https://github.com/gravitational/teleport/blob/72825b1b9e1cc47a5c5468a910258405e4b238b3/api/types/resource_153.go#L32), specifically:

> `Embedding or further extending this interface is highly discouraged.`

... and this was a way to 

- avoid expanding an interface when explicitly advised not to (the only place `Resource153` is mentioned is the type constraint on `UnifiedResource153`),
- avoid inadvertently changing the behaviour of existing values wrapped by `resource153ToLegacyAdapter`, and
- stop any attempt to wrap a resource type without a `Clone()` method at compile time. 

My gut feeling is that the "correct" thing to do is probably slim down the [`resource`](https://github.com/gravitational/teleport/blob/72825b1b9e1cc47a5c5468a910258405e4b238b3/lib/services/unified_resource.go#L723) interface used in the Resource Cache as per the `Resource153` godoc: 

> If you do need to represent resources in a generic manner, consider
> declaring a smaller interface with only what you need.

But that refactor is well outside the scope of this PR. 

Anyway, that was my reasoning for things being as they are. If expanding the `types.Resource153` interface (and the accompanying `resource153ToLegacyAdapter` etc) is the lesser of two evils, I'm happy to go that way.

File: lib/services/unified_resource_adapter.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+// UnifiedResource153Adapter is a wrapper around a newer, RFD153-style resource
+// that provides the newer style resources with the interfaces required for use
+// with the Unified resource cache.
+type UnifiedResource153Adapter[T UnifiedResource153[T]] struct {
```
Comment (smallinsky) [inline lib/services/unified_resource_adapter.go:23]: The UnifiedResource153Adapter duplicates 80% of the functionality in [Resource153ToLegacy](https://github.com/gravitational/teleport/blob/master/api/types/resource_153.go#L123).

The only missing methods required to make the Unified Resource fit the ResourceWithLabels interface are:

```
func (r Resource) Origin() string {
}

func (r Resource) SetOrigin(s string) {
}

func (r Resource) GetLabel(key string) (value string, ok bool) {
}

func (r Resource) GetAllLabels() map[string]string {
}

func (r Resource) GetStaticLabels() map[string]string {
}

func (r Resource) SetStaticLabels(sl map[string]string) {
}

func (r Resource) MatchSearch(searchValues []string) bool {
}
```


Would it be feasible to simply add these missing methods to [identitycenter.go](https://github.com/gravitational/teleport/blob/master/lib/services/identitycenter.go#L36), or can the existing [Resource153ToLegacy](https://github.com/gravitational/teleport/blob/master/api/types/resource_153.go#L123) implementation be leveraged to avoid introducing additional abstractions and new interfaces for Resource153 and legacy resources?

File: lib/web/ui/app.go [hunk 9]
```diff
@@ -155,6 +172,13 @@ func MakeApp(app types.Application, c MakeAppsConfig) App {
 	return resultApp
 }
 
+func mapPermissionSet(ps *types.IdentityCenterPermissionSet) IdentityCenterPermissionSet {
```
Comment (smallinsky) [inline lib/web/ui/app.go:175]: Do we really need a separate function for type conversion if it is only used once in the code?
The map... naming is misaligning suggesting that there map... function does some kine of logic but this is just simple conversion.

File: lib/auth/auth_with_roles.go [hunk 10]
```diff
@@ -1822,9 +1832,12 @@ func (r resourceChecker) CanAccess(resource types.Resource) error {
...
+		return r.CheckAccess(rr, state)
```
Comment (smallinsky) [inline lib/auth/auth_with_roles.go:1837]: Is there any UT that covers this flow if not can we add test coverage for the IdentityCenterAccount CheckAccess flow.

File: lib/services/unified_resource.go [hunk 11]
```diff
@@ -470,6 +480,7 @@ func (c *UnifiedResourceCache) getResourcesAndUpdateCurrent(ctx context.Context)
 	putResources[types.KubeServer](c, newKubes)
 	putResources[types.SAMLIdPServiceProvider](c, newSAMLApps)
 	putResources[types.WindowsDesktop](c, newDesktops)
+	putResources[UnifiedResource153Adapter[IdentityCenterAccount]](c, newICAccounts)
```
[the remark this answers is not in the document]
Comment (tcsc) [reply lib/services/unified_resource.go]: Accounts gets converted to an App on read

File: lib/auth/auth_with_roles.go (hunk 10 above)
Comment (tcsc) [reply lib/auth/auth_with_roles.go]: Added in e055c04

File: lib/services/unified_resource.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (tcsc) [reply lib/services/unified_resource.go]: Mainly so they are governed by the RBAC conditions for Identity Center resources, rather than app resources.

**Review by rosstimothy — COMMENTED** [review]
I'd suggest splitting this up into more focused PRs. There is a lot of what feels like glued on functionality that isn't directly related to unified resource caching being included here.

File: api/proto/teleport/legacy/types/types.proto [hunk 12]
```diff
@@ -3416,6 +3420,18 @@ message RoleConditions {
...
+  // Identity Center resource
+  wrappers.LabelValues AccountAssignmentLabels = 44 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "account_assignment_labels,omitempty",
+    (gogoproto.customtype) = "Labels"
+  ];
+
+  // AccountAssignmentLabels is a label expression used to match any sort of
+  // Identity Center resource
+  string AccountAssignmentLabelsExpression = 45 [(gogoproto.jsontag) = "account_assignment_labels_expression,omitempty"];
```
[the remark this answers is not in the document]
Comment (r0mant) [reply api/proto/teleport/legacy/types/types.proto:3434]: Yes, didn't we agree before that we don't want to require users to set another set of labels in their roles. We should use AccountAssignments only.

File: api/proto/teleport/legacy/types/types.proto [hunk 13]
```diff
@@ -3416,6 +3420,18 @@ message RoleConditions {
     (gogoproto.nullable) = false,
     (gogoproto.jsontag) = "github_permissions,omitempty"
   ];
+
+  // AccountAssignmentLabels is the label set used to match any sort of
+  // Identity Center resource
+  wrappers.LabelValues AccountAssignmentLabels = 44 [
+    (gogoproto.nullable) = false,
+    (gogoproto.jsontag) = "account_assignment_labels,omitempty",
```
Comment (flyinghermit) [inline api/proto/teleport/legacy/types/types.proto:3428]: Two concerns on label name `account_assignment_labels`:
1. This label will be obsolete/ or we need to introduce another label once again when we support requesting access to both AWS account and AWS application.
2. `account_assignment_labels` is generic enough and might collide labels we might need for future integrations such as GCP accounts, Azure accounts etc (if we ever end up implementing them).  

Can we get away with `aws_ic_assigment_labels` that can be applied to both AWS account and application?

File: lib/services/unified_resource.go [hunk 14]
```diff
@@ -879,13 +910,78 @@ func MakePaginatedResource(ctx context.Context, requestType string, r types.Reso
...
+						SubKind: types.AppSubKindIdentityCenterAccount,
+						Version: types.V3,
+						Metadata: types.Metadata{
+							Name:        acct.Spec.Name,
+							Description: acct.Spec.Description,
+							Labels:      maps.Clone(acct.Metadata.Labels),
+						},
+						Spec: types.AppSpecV3{
+							URI:        apiconstants.AWSConsoleURL,
+							PublicAddr: apiconstants.AWSConsoleURL,
```
Comment (flyinghermit) [inline lib/services/unified_resource.go:966]: The value of `apiconstants.AWSConsoleURL` is https://console.aws.amazon.com, which sends user to the root AWS console that lets them pick account with permission set. Since we want to launch users directly to AWS console with  specific account+permission set, we need URL value that contains Identity Center instance ID and account ID.

In the POC, I used `StartUrl` field to contain such value https://github.com/gravitational/teleport.e/blob/sshah/idc-dev/lib/aws/identitycenter/accounts.go#L156. But in our current version, we have the [StartUrl field defined](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/identitycenter/v1/identitycenter.proto#L67C1-L68C24) but we missed to enrich the field with instance ID and account ID on[ initial account import](https://github.com/gravitational/teleport.e/blob/master/lib/aws/identitycenter/accounts.go#L91C1-L110C2). 

tldr; We need to update the [newIdentityCenterAccount](https://github.com/gravitational/teleport.e/blob/master/lib/aws/identitycenter/accounts.go#L91C1-L110C2) func to enrich StartUrl field and the use the value here.

File: lib/services/unified_resource.go [hunk 15]
```diff
@@ -879,13 +910,78 @@ func MakePaginatedResource(ctx context.Context, requestType string, r types.Reso
...
+							AWS: &types.AppAWS{
+								ExternalID: acct.Spec.Id,
+							},
+							IdentityCenter: &types.AppIdentityCenter{
+								AccountID:      acct.Spec.Id,
+								PermissionSets: pss,
```
Comment (flyinghermit) [inline lib/services/unified_resource.go:972]: Given your another PR https://github.com/gravitational/teleport/pull/49580 and that we use account assignment resource list to let user select desired permission set for access request, why do we need to include `PermissionSets` field and its value?

File: lib/services/unified_resource.go (hunk 14 above)
Comment (tcsc) [reply lib/services/unified_resource.go:966]: Updated the references here. Will update the account constructor in a later PR.

File: api/types/resource_153.go [hunk 16]
```diff
@@ -181,34 +185,83 @@ func (r *resource153ToLegacyAdapter) GetMetadata() Metadata {
...
+func (r *Resource153ToLegacyAdapter) MatchSearch(searchValues []string) bool {
+	fieldVals := append(utils.MapToStrings(r.GetAllLabels()), r.GetName())
+	return MatchSearch(fieldVals, searchValues, nil)
+}
+
+func (r *Resource153ToLegacyAdapter) CloneResource() ResourceWithLabels {
+	if cloner, ok := r.inner.(interface{ CloneResource() Resource153 }); ok {
+		return &Resource153ToLegacyAdapter{inner: cloner.CloneResource().(Resource153)}
+	}
+	panic("interface Resource153 does not implement CloneResource for the wrapped type")
```
[the remark this answers is not in the document]
Comment (tcsc) [reply api/types/resource_153.go:266]: I'm not sure what the correct alternative is for resources that don't implement CloneResource, short of `reflect.DeepCopy`.

**Comment by tcsc** [comment]
Label matching removed and replaces with custom role matchers.

**Review by smallinsky — COMMENTED** [review]
I have left some comment/suggestion

Also  and wonder about: 
https://github.com/gravitational/teleport/pull/49301#discussion_r1869574315 

> it feels like it’s attempting to introduce multiple changes at once—such as RBAC > checks, backend updates, and adding new resources.
> 
> To make the review process more efficient and focused it might be helpful to break > these changes into separate smaller PRs his way each part can be reviewed and > discussed independently.
>
> For example - the PR description mentions: Exposes Identity Center accounts as Apps > in Unified Resource Cache but it’s not immediately clear how or why RBAC alignment > backend changes ** adding new resources** are tied to this goal.
> 
> Splitting the changes could make the purpose of each adjustment clearer and help reviewers better understand the overall intent.

I think that this is worth to consider as it could significantly improve the review process and the overall quality of the PR.

File: api/types/resource_153.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (tcsc) [reply api/types/resource_153.go:285]: This is why I had a separate implementation in the first place.

File: api/types/app.go [hunk 17]
```diff
@@ -456,6 +458,23 @@ func (a *AppV3) checkTCPPorts() error {
...
 
+// GetIdentityCenter returns the Identity Center information for the app, if any.
+// May be nil.
+func (a *AppV3) GetIdentityCenter() *AppIdentityCenter {
+	return a.Spec.IdentityCenter
+}
+
+// GetDisplayName fetches a human-readable display name for the App.
+func (a *AppV3) GetDisplayName() string {
+	// Only Identity Center apps have a display name at this point. Returning
```
Comment (r0mant) [inline api/types/app.go:469]: Technically Okta apps also have a different display name because their actual name is an internal Okta ID. What do we use for them?

File: lib/auth/auth_with_roles.go [hunk 18]
```diff
@@ -145,6 +145,26 @@ func (a *ServerWithRoles) authConnectorAction(resource string, verb string) erro
...
+	return a.identityCenterActionNamespace(apidefaults.Namespace, resource, verbs...)
+}
+
+// identityCenterActionNamespace is a special checker that grants access to
+// Identity Center resources. In order to simplify the writing of role condition
+// statements, the various Identity Center resources are bundled up under an
+// umbrella `KindIdentityCenter` resource kind. This means that if access to the
+// target resource is not explicitly denied, then the user has a second chance
+// to get access via the generic resource kind.
+func (a *ServerWithRoles) identityCenterActionNamespace(namespace string, resource string, verbs ...string) error {
```
Comment (r0mant) [inline lib/auth/auth_with_roles.go:160]: Nit: We literally never use a namespace other than "default", do we really need a separate `identityCenterActionNamespace` method?

I would consider just keeping `identityCenterAction` and hardcoding `apidefaults.Namespace` to keep things a little simpler.

File: lib/services/identitycenter.go [hunk 19]
```diff
@@ -226,3 +236,84 @@ type IdentityCenter interface {
...
+	return &IdentityCenterMatcher{
+		accountID:        account.GetSpec().GetId(),
+		permissionSetARN: nil,
+	}
+}
+
+// NewIdentityCenterAccountAssignmentMatcher creates a new [IdentityCenterMatcher]
+// configured to match the supplied [IdentityCenterAccountAssignment].
+func NewIdentityCenterAccountAssignmentMatcher(account IdentityCenterAccountAssignment) *IdentityCenterMatcher {
+	psARN := account.GetSpec().GetPermissionSet().GetArn()
```
Comment (r0mant) [inline lib/services/identitycenter.go:252]: Can any of these be nil?

File: lib/services/identitycenter.go [hunk 20]
```diff
@@ -226,3 +236,84 @@ type IdentityCenter interface {
...
+		}
+
+		if psMatches {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (m *IdentityCenterMatcher) matchExpression(target, accountExpression string) (bool, error) {
```
Comment (r0mant) [inline lib/services/identitycenter.go:300]: Nit: This is used to match both accounts and permission sets from what I can see.
```suggestion
func (m *IdentityCenterMatcher) matchExpression(target, expression string) (bool, error) {
```

File: lib/services/identitycenter.go [hunk 21]
```diff
@@ -226,3 +236,84 @@ type IdentityCenter interface {
...
+// an Account.
+type IdentityCenterMatcher struct {
+	accountID        string
+	permissionSetARN *string
+}
+
+// Match implements RoleMatcher for IdentityCenterMatcher. It attempts to match
+// the Account Assignments in a Role Condition against a known Account ID or
+// (Account ID, PermissionSetARN) pair
+func (m *IdentityCenterMatcher) Match(role types.Role, condition types.RoleConditionType) (bool, error) {
```
Comment (r0mant) [inline lib/services/identitycenter.go:273]: We should probably add a TODO note somewhere to add support for templating variables. I presume people would want to propagate those from their IdP via `{{external.account_assignments}}` for example or something similar.

File: lib/services/role.go [hunk 22]
```diff
@@ -2589,18 +2595,18 @@ func (set RoleSet) checkAccess(r AccessCheckable, traits wrappers.Traits, state
...
-		matchLabels, labelsMessage, err := checkRoleLabelsMatch(types.Deny, role, traits, r, isDebugEnabled)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-		if matchLabels {
-			debugf("Access to %v %q denied, deny rule in role %q matched; match(namespace=%v, %s)",
-				r.GetKind(), r.GetName(), role.GetName(), namespaceMessage, labelsMessage)
-			return trace.AccessDenied("access to %v denied. User does not have permissions. %v",
-				r.GetKind(), additionalDeniedMessage)
+		if requiresLabelMatching {
```
Comment (r0mant) [inline lib/services/role.go:2598]: Here and below, let's add an explicit log entry that would say the label matching is being skipped for resource X or something along those lines.

File: lib/services/role.go [hunk 23]
```diff
@@ -2589,18 +2595,18 @@ func (set RoleSet) checkAccess(r AccessCheckable, traits wrappers.Traits, state
...
+			matchLabels, labelsMessage, err := checkRoleLabelsMatch(types.Deny, role, traits, r, isDebugEnabled)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			if matchLabels {
+				debugf("Access to %v %q denied, deny rule in role %q matched; match(namespace=%v, %s)",
+					r.GetKind(), r.GetName(), role.GetName(), namespaceMessage, labelsMessage)
+				return trace.AccessDenied("access to %v denied. User does not have permissions. %v",
+					r.GetKind(), additionalDeniedMessage)
+			}
```
[the remark this answers is not in the document]
Comment (r0mant) [reply lib/services/role.go:2608]: This has been here before though and I think "allow" and "deny" code paths are slightly different so may be tricky to factor out. I think it's ok to keep as-is.

**Review by r0mant — APPROVED** [review]
This makes much more sense than the previous iteration, lgtm once remaining feedback has been addressed.

File: lib/web/ui/app.go [hunk 24]
```diff
@@ -76,6 +82,17 @@ type UserGroupAndDescription struct {
...
 
+// IdentityCenterPermissionSet holds information about Identity Center
+// Permission Sets for transmission to the UI
+type IdentityCenterPermissionSet struct {
+	Name string `json:"name"`
+	ARN  string `json:"arn"`
+	// AssignmentName is the assignment resource that will provision an Account
+	// Assignment for this Permission Set on the enclosing account
+	AssignmentName  string `json:"accountAssignment,omitempty"`
+	RequiresRequest bool   `json:"requiresRequest,omitempty"`
```
[the remark this answers is not in the document]
Comment (tcsc) [reply lib/web/ui/app.go:93]: This will be used when we filter the resources based on access.

**Comment by tcsc** [comment]
Moved RBAC in to another PR

File: lib/services/unified_resource.go [hunk 25]
```diff
@@ -879,13 +913,78 @@ func MakePaginatedResource(ctx context.Context, requestType string, r types.Reso
...
+						SubKind: types.KindIdentityCenterAccount,
+						Version: types.V3,
+						Metadata: types.Metadata{
+							Name:        acct.Spec.Name,
+							Description: acct.Spec.Description,
+							Labels:      maps.Clone(acct.Metadata.Labels),
+						},
+						Spec: types.AppSpecV3{
+							URI:        acct.Spec.StartUrl,
+							PublicAddr: acct.Spec.StartUrl,
```
[the remark this answers is not in the document]
Comment (tcsc) [reply lib/services/unified_resource.go:988]: This is coming in a change discussed with @flyinghermit

File: lib/web/ui/app.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (smallinsky) [reply lib/web/ui/app.go:95]: Could you add a go docs ?

**Review by smallinsky — APPROVED** [review]
